django-server/server/api/views.py

```diff
@@ -112,6 +112,131 @@ def get_queryset(self):
+            curr_overrides = curr_range.filter(event_type=CameraEvent.EventType.OCCUPANCY_OVERRIDE.value)
+            if len(curr_overrides) == 0:
+                people_inside += self.count_people(curr_range)
+            else:
+                last_override = curr_overrides.latest('timestamp')
+                people_inside = last_override + self.count_people(curr_range
+                                                                  .filter(timestamp__gt=last_override.timestamp))
+            begin += self.get_timestamp_delta(begin)
+            end += self.get_timestamp_delta(begin)
+            results.append((people_inside, begin))
+            if finish_now:
+                break
+
+        occupancy = Cafeteria.objects.get(id=id).occupancy
+        return [OccupancyStatsData(id=index, timestamp_name=self.get_timestamp_name(stamp), occupancy=inside,
+                                   occupancy_relative=min(float(inside) / float(occupancy), 1.0))
+                for index, (inside, stamp) in enumerate(results)]
+
+    def count_people(self, queryset):
+        counter = 0
+        for event in queryset:
+            if event.event_type == CameraEvent.EventType.PERSON_ENTERED.value:
+                counter += 1
+            elif event.event_type == CameraEvent.EventType.PERSON_LEFT.value:
+                counter -= 1
+        return counter
+
+    def filter_queryset(self, queryset):
+        return queryset
+
+    @abc.abstractmethod
+    def get_timestamp_delta(self, datetime: datetime):
+        pass
+
+    @abc.abstractmethod
+    def get_timestamp_name(self, datetime: datetime):
+        pass
+
+
+class HourStatsView(StatsView):
+
+    def get_timestamp_delta(self, datetime: datetime):
+        return timedelta(hours=1)
+
+    def get_timestamp_name(self, datetime: datetime):
+        return '{}'.format(datetime.hour)
+
+
+class DayStatsView(StatsView):
+
+    def get_timestamp_delta(self, datetime: datetime):
+        return timedelta(days=1)
+
+    def get_timestamp_name(self, datetime: datetime):
+        week_days = ['Mon', 'Tue', 'Wed', 'Thr', 'Fri', 'Sat', 'Sun']
+        return week_days[datetime.weekday()]
+
+
+class WeekStatsView(StatsView):
+
+    def get_timestamp_delta(self, datetime: datetime):
+        return timedelta(weeks=1)
+
+    def get_timestamp_name(self, datetime: datetime):
+        return '{} week'.format(datetime.isocalendar()[1])
+
+
+class MonthStatsView(StatsView):
+
+    def get_timestamp_delta(self, datetime: datetime):
+        return timedelta(days=monthrange(datetime.year, datetime.month)[1])
+
+    def get_timestamp_name(self, datetime: datetime):
+        return '{}'.format(datetime.day)
+
+
+class YearStatsView(StatsView):
+
+    def get_timestamp_delta(self, datetime: datetime):
+        return timedelta(days=(datetime.replace(year=datetime.year + 1) - datetime).days)
+
+    def get_timestamp_name(self, datetime: datetime):
+        return '{}'.format(datetime.year)
+
```

Comment:
See earlier comment about copying.

Reply:
this is done only once to specify the ranges of filter type so I leave it there